### PR TITLE
feat(compiler): Support MetaProperty

### PR DESCRIPTION
**Pull request by nikeee**
## Summary
See #29737

## How did you test this change?
As the feature requires module support and the test runner does currently not support running tests as modules, I could only test it via playground.

### CODE REVIEW

**Review by josephsavona — APPROVED** [review]
looks great, thank you!

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts [hunk 1]
```diff
@@ -1204,6 +1204,7 @@ function inferBlock(
         };
         break;
       }
+      case "MetaProperty":
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts:1207]: hmmm this isn't safe: the `new.target` meta-property may refer to a mutable object (constructor function), so we can't consider this value to be immutable. If we want to do this, we'll need to restrict support to just `import` meta properties which shouldn't be modified

**Review by josephsavona — CHANGES_REQUESTED** [review]
Sorry, i spoke too soon. I just double-checked what other meta properties there are and realized it isn't guaranteed to be an immutable object.

File: compiler/packages/babel-plugin-react-compiler/src/HIR/BuildHIR.ts [hunk 2]
```diff
@@ -2427,6 +2427,15 @@ function lowerExpression(
...
     }
+    case "MetaProperty": {
+      let expr = exprPath as NodePath<t.MetaProperty>;
+      return {
+        kind: "MetaProperty",
+        meta: expr.node.meta.name,
+        property: expr.node.property.name,
+        loc: expr.node.loc ?? GeneratedSource,
+      };
+    }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/BuildHIR.ts:2438]: i think what we can do is limit this to `import` meta properties for now, and throw a Todo error for other `meta` values.

Comment (nikeee) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/BuildHIR.ts]: I constrained it to `expr.node.meta.name === "import" && expr.node.property.name === "meta"`

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts (hunk 1 above)
Comment (nikeee) [reply compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts:1207]: Great feedback, I'll look into this.

File: compiler/packages/babel-plugin-react-compiler/src/HIR/BuildHIR.ts [hunk 3]
```diff
@@ -2427,6 +2427,21 @@ function lowerExpression(
...
+        expr.node.property.name === "meta"
+      ) {
+        return {
+          kind: "MetaProperty",
+          meta: expr.node.meta.name,
+          property: expr.node.property.name,
+          loc: expr.node.loc ?? GeneratedSource,
+        };
+      }
+      return { kind: "UnsupportedNode", node: exprNode, loc: exprLoc };
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/BuildHIR.ts:2443]: Note that it’s not enough to return an unsupported node, we also have to actually record a todo error for this case. 

Please also add a fixture using `new.target` to demonstrate that we reject it. You can name the fixture with “error.todo-new-target-meta-property.js” - the “error” prefix tells our test runner to fail unless the fixture causes an error. And the “todo-“ is a reminder to us that the input should someday be supported.

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts (hunk 1 above)
Comment (nikeee) [reply compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts]: I tried addressing this in 3e3b0174c2 and I'm not sure if I correctly handled the case when the property is no from `import.meta`. Can you look into that?

Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts]: Awesome! On my phone but this looks right, will do a proper review next week

**Review by josephsavona — COMMENTED** [review]
this is great!

**Review by josephsavona — APPROVED** [review]

**Pull request merged**